---
title: "Get GO & TF Annotations"
date: 2020-11-25
output: 
  bookdown::html_document2:    
  fig_caption: yes
---

```{r}
p_load(biomaRt)
listMarts()

ensembl <- useMart("ensembl")
ensembl <- useDataset("hsapiens_gene_ensembl", mart = ensembl)

listDatasets(ensembl)
listAttributes(ensembl) %>% filter(! name %>% str_detect("homolog"))

searchAttributes(mart = ensembl, pattern = "GO")

go_attributes <- searchAttributes(mart = ensembl, pattern = "GO ") %>% pull(name) %>% .[-4]

getwd()
p_load(knitr)
```

## Read transcription factor databases

From <http://bioinfo.life.hust.edu.cn/AnimalTFDB/#!/>

```{r}
tfs_db <- read_table2(file = 'http://bioinfo.life.hust.edu.cn/static/AnimalTFDB3/download/Homo_sapiens_TF') %>% 
            set_names(tolower(names(.))) %>% 
              rename(hgnc_symbol = "symbol") %>% 
                dplyr::select(-id)
# 1,665 TFs on 2020-11-25

## Transcription cofactors
tf_cofactors_db <- read_tsv("http://bioinfo.life.hust.edu.cn/static/AnimalTFDB3/download/Homo_sapiens_TF_cofactors") %>% 
                      set_names(tolower(names(.))) %>%
                        rename(hgnc_symbol = "symbol")
# 1,025 TF cofactors on 2020-11-25
```

# Get GO annotations

for all cell type-area marker genes

```{r fxn_getGO}
getGO <- function(genes = x) {
  
              getBM(attributes = c('hgnc_symbol', 'description', go_attributes), 
                    filters = 'hgnc_symbol', 
                    values = genes, 
                    mart = ensembl) %>%
    
              filter(! namespace_1003 %>% str_detect("cellular_component") &
                     ! name_1006 == '') %>% 
                add_count(name_1006) %>% arrange(desc(n), name_1006) %>% 
                  # mutate(name_1006 = as_factor(name_1006)) %>%
                    dplyr::select(hgnc_symbol, name_1006, n, description, everything())
}


getGO_map <- function(groups_tbl, celltype) {
    
    groups_tbl %>% 
    # dplyr::filter(item2 %>% str_detect(celltype)) %>% 
                dplyr::select(item2, genes) %>% 
                  pmap(~ tibble(group = ..1, 
                                go_df = list(getGO(genes = ..2))
                                )
                       ) %>%
                    reduce(bind_rows)
}
```

## . . . . . . . . . . . . . . . . . . . . . . . . . . . .

## Radial glia: PFC: all marker genes

```{r}
# x <- pairs$by_area %>% filter(item2 == "rg_pfc_all") %>% pull(genes) %>% .[[1]]

goids_x <- getGO(genes = x)
```

## Genes with expression-related annotations

```{r}
getGenesWithExpressionGO <- function(go_df) {
  
    group_markers_n <- go_df$hgnc_symbol %>% unique %>% length
    
    expression_go <- go_df %>% filter(name_1006 %>% str_detect("transcription|enhancer|methylation|expression") & 
                                      ! name_1006 %>% str_detect("viral"))
    
    expression_go <- tibble(# group = "rg_pfc_all", 
                           
                           expression_go_genes =  expression_go$hgnc_symbol %>% 
                                                    unique %>% paste(collapse= ";"),
                           expression_go_genes_n = expression_go$hgnc_symbol %>% 
                                                    unique %>% length,
                           
                           expression_go_genes_frac = (expression_go_genes_n / group_markers_n) %>% round(2),
                           
                           expression_go_df = list(expression_go)
                         
                    )
}
```

## Intersect with transcription factor and co-factor DBs.

```{r}
getGenesWithTFannotation <- function(go_df) {
  
  group_markers_n <- go_df$hgnc_symbol %>% unique %>% length

  tf_go_df <- list(left_join(go_df, tfs_db) %>% 
                      filter(! is.na(family))
                  )
  
  tf_cofactor_go_df <- list(left_join(go_df, tf_cofactors_db) %>% 
                               filter(! is.na(family))
                )
  
  tfs <- tibble(# group = "rg_pfc_all", 
                group_markers_n = group_markers_n,
                
                tf_genes =  tf_go_df[[1]]$hgnc_symbol %>% unique %>% paste(collapse= ";"),
                tf_genes_n = tf_go_df[[1]]$hgnc_symbol %>% unique %>% length,
                tf_genes_frac = (tf_genes_n / group_markers_n) %>% round(2),
                tf_go_df = tf_go_df
          )
  
  tf_cofactors <- tibble(# group = "rg_pfc_all", 
                         
                         tf_cofactor_genes = tf_cofactor_go_df[[1]]$hgnc_symbol %>% 
                                                  unique %>% paste(collapse= ";"),
                         
                         tf_cofactor_genes_n = tf_cofactor_go_df[[1]]$hgnc_symbol %>% 
                                                  unique %>% length,
                         tf_cofactor_genes_frac = (tf_cofactor_genes_n / group_markers_n) %>% round(2),
                         
                         tf_cofactor_go_df = tf_cofactor_go_df
                       
                  )
  
  tfs_and_cofactors <- bind_cols(tfs, tf_cofactors)
  
}

 %>% left_join(expression_go) %>% add_column(all_markers_go_df = list(goids_x))
```