---
title: "Analysis of areal markers"
output: html_notebook
---


```{r}
remotes::install_github("federicomarini/GeneTonic", 
                        dependencies = TRUE, build_vignettes = TRUE)

BiocManager::install("clusterProfiler")
BiocManager::install("pathview")
BiocManager::install("enrichplot")
library(clusterProfiler)
library(enrichplot)
```

```{r}
genes <- pairs.norm %>% filter(item2 == "rg_pfc_unique") %>% .$genes %>% .[[1]]

```

```{r}
p_load(biomaRt)

listMarts()

ensembl <- useMart("ensembl")
ensembl <- useDataset("hsapiens_gene_ensembl", mart=ensembl)

listDatasets(ensembl)

listAttributes(ensembl) %>% filter(! name %>% str_detect("homolog"))

searchAttributes(mart = ensembl, pattern = "GO|")
```

```{r}

goids <- getBM(attributes = c('hgnc_symbol', 'description', 'name_1006', 'definition_1006', 'go_id','namespace_1003'), 
              filters = 'hgnc_symbol', 
              values = genes, 
              mart = ensembl)

goids %>% as_tibble %>% dplyr::select(hgnc_symbol, description, name_1006, definition_1006, go_id, namespace_1003) %>% filter(namespace_1003 %in% c("molecular_function", "biological_process"))

goids %>% count(name_1006) %>% arrange(desc(n))
```
```{r}
goids %>% filter(name_1006 %>% str_detect("transcription")) %>% distinct(hgnc_symbol, name_1006, .keep_all = TRUE)
```
Categories with the word 'transcription'
```{r}
goids %>% filter(name_1006 %>% str_detect("transcription")) %>% distinct(hgnc_symbol, name_1006, .keep_all = TRUE) %>% count(name_1006)
```

```{r}
goids %>% filter(name_1006 %>% str_detect("transcription")) %>% distinct(hgnc_symbol, name_1006, .keep_all = TRUE) %>% 
  ggplot() + geom_tile(aes(y = name_1006, x = as_factor(hgnc_symbol), fill = name_1006))
```
```{r}
x <- pairs.norm %>% filter(item2 == "rg_pfc_all") %>% .$genes %>% .[[1]]

goids_x <- getBM(attributes = c('hgnc_symbol', 'description', 'name_1006', 'definition_1006', 'go_id','namespace_1003'), 
              filters = 'hgnc_symbol', 
              values = x, 
              mart = ensembl)

goids_x %>% filter(name_1006 %>% str_detect("transcription|enhancer|methylation") & ! name_1006 %>% str_detect("viral")) %>% distinct(hgnc_symbol, name_1006, .keep_all = TRUE) %>% add_count(name_1006) %>% arrange(desc(n)) %>% distinct(hgnc_symbol, .keep_all = TRUE) %>% dplyr::select(hgnc_symbol, name_1006, n, everything())
# 149 genes are classified as some sort of transcription GO
%>% 
  ggplot() + geom_tile(aes(y = name_1006, x = as_factor(hgnc_symbol), fill = name_1006)) + theme(legend.position = "none", axis.text.x = element_text(angle = 90))
```

```{r}
tfs <- read_table2("animalDB/Homo_sapiens_TF.txt") %>% set_names(tolower(names(.)))

left_join(x %>% tibble(symbol = .), tfs) %>% filter(! is.na(family))

# 43 genes are in AnimalDB
```


```{r}
tf_cofactors <- read_tsv("animalDB/Homo_sapiens_TF_cofactors.txt") %>% set_names(tolower(names(.)))

left_join(x %>% tibble(symbol = .), tfs) %>% filter(! is.na(family))

# 43 genes are in AnimalDB
```


