---
title: "Analysis of areal markers: Transcription factors"
date: "2020-11-12"
output: 
  bookdown::html_document2:
    fig_caption: yes
---

```{r load-libraries}
p_load(tidyverse)
p_load(patchwork)
p_load(janitor)
p_load(magrittr)
p_load(Seurat)
p_load(plotly)


library(clusterProfiler)
library(enrichplot)
p_load(rrvgo)
```

```{r}
remotes::install_github("federicomarini/GeneTonic", 
                        dependencies = TRUE, build_vignettes = TRUE)

BiocManager::install("clusterProfiler")
BiocManager::install("pathview")
BiocManager::install("enrichplot")
```

```{r}
# save.image(file = "R/2020-11-15_transcription_factor_composition_workspace.RData")
load(file = "R/2020-11-15_transcription_factor_composition_workspace.RData")
sms("Loaded image")
```

# Read pair-wise gene marker intersection lists (from Sankey plots.)

```{r read-data}
pairs <- 
  list(by_area_stage = read_rds("../../sankey_plots/out_intermediate/sankey_markers_pairs_exnLineage.rds"),
       by_area = read_rds("../../sankey_plots/out_intermediate/sankey_markers_pairs_exnLineage_byArea_only.rds")
          )
```

# Select biomaRt database and attributes to fetch {#select-biomart-database-and-attributes-to-fetch}

```{r}
p_load(biomaRt)

listMarts()

ensembl <- useMart("ensembl")
ensembl <- useDataset("hsapiens_gene_ensembl", mart = ensembl)

listDatasets(ensembl)
listAttributes(ensembl) %>% filter(! name %>% str_detect("homolog"))

searchAttributes(mart = ensembl, pattern = "GO")

go_attributes <- searchAttributes(mart = ensembl, pattern = "GO ") %>% pull(name) %>% .[-4]
```

# Read transcription factor databases

```{r}
tfs_db <- read_table2("animalDB/Homo_sapiens_TF.txt") %>% set_names(tolower(names(.)))
tfs_db %<>% rename(hgnc_symbol = "symbol") %>% select(-id)
# 1,665 TFs

## Transcription cofactors
tf_cofactors_db <- read_tsv("animalDB/Homo_sapiens_TF_cofactors.txt") %>% set_names(tolower(names(.)))
tf_cofactors_db %<>% rename(hgnc_symbol = "symbol")
# 1,025 TF cofactors
```

# Get GO annotations

for all cell type-area marker genes

```{r fxn:getGO}
getGO <- function(genes = x) {
  
              getBM(attributes = c('hgnc_symbol', 'description', go_attributes), 
                    filters = 'hgnc_symbol', 
                    values = genes, 
                    mart = ensembl) %>%
    
              filter(! namespace_1003 %>% str_detect("cellular_component") &
                     ! name_1006 == '') %>% 
                add_count(name_1006) %>% arrange(desc(n), name_1006) %>% 
                  # mutate(name_1006 = as_factor(name_1006)) %>%
                    dplyr::select(hgnc_symbol, name_1006, n, description, everything())
}


getGO_map <- function(groups_tbl, celltype) {
    
    groups_tbl %>% 
    # dplyr::filter(item2 %>% str_detect(celltype)) %>% 
                dplyr::select(item2, genes) %>% 
                  pmap(~ tibble(group = ..1, 
                                go_df = list(getGO(genes = ..2))
                                )
                       ) %>%
                    reduce(bind_rows)
}
```

## Radial glia: PFC: unique marker genes

```{r}
genes <- pairs$by_area %>% filter(item2 == "rg_pfc_unique") %>% .$genes %>% .[[1]]
```

goids %\>% as_tibble %\>% dplyr::select(hgnc_symbol, description, name_1006, definition_1006, go_id, namespace_1003) %\>% filter(namespace_1003 %in% c("molecular_function", "biological_process"))

goids %\>% count(name_1006) %\>% arrange(desc(n))

```{r}
goids %>% filter(name_1006 %>% str_detect("transcription")) %>% distinct(hgnc_symbol, name_1006, .keep_all = TRUE)
```

Categories with the word 'transcription'

```{r}
goids %>% filter(name_1006 %>% str_detect("transcription")) %>% distinct(hgnc_symbol, name_1006, .keep_all = TRUE) %>% count(name_1006)
```

```{r}
goids %>% filter(name_1006 %>% str_detect("transcription")) %>% distinct(hgnc_symbol, name_1006, .keep_all = TRUE) %>% 
  ggplot() + geom_tile(aes(y = name_1006, x = as_factor(hgnc_symbol), fill = name_1006))
```

## . . . . . . . . . . . . . . . . . . . . . . . . . . . .

## Radial glia: PFC: all marker genes

```{r}
# x <- pairs$by_area %>% filter(item2 == "rg_pfc_all") %>% pull(genes) %>% .[[1]]

goids_x <- getGO(genes = x)
```

## Genes with expression-related annotations

```{r}
getGenesWithExpressionGO <- function(go_df) {
  
    group_markers_n <- go_df$hgnc_symbol %>% unique %>% length
    
    expression_go <- go_df %>% filter(name_1006 %>% str_detect("transcription|enhancer|methylation|expression") & 
                                      ! name_1006 %>% str_detect("viral"))
    
    expression_go <- tibble(# group = "rg_pfc_all", 
                           
                           expression_go_genes =  expression_go$hgnc_symbol %>% 
                                                    unique %>% paste(collapse= ";"),
                           expression_go_genes_n = expression_go$hgnc_symbol %>% 
                                                    unique %>% length,
                           
                           expression_go_genes_frac = (expression_go_genes_n / group_markers_n) %>% round(2),
                           
                           expression_go_df = list(expression_go)
                         
                    )
}
```

## Intersect with transcription factor and co-factor DBs.

```{r}
getGenesWithTFannotation <- function(go_df) {
  
  group_markers_n <- go_df$hgnc_symbol %>% unique %>% length

  tf_go_df <- list(left_join(go_df, tfs_db %>% dplyr::select(hgnc_symbol, family)) %>% 
                      filter(! is.na(family))
                  )
  
  tf_cofactor_go_df <- list(left_join(go_df, tf_cofactors_db) %>% 
                               filter(! is.na(family))
                )
  
  tfs <- tibble(# group = "rg_pfc_all", 
                group_markers_n = group_markers_n,
                
                tf_genes =  tf_go_df[[1]]$hgnc_symbol %>% unique %>% paste(collapse= ";"),
                tf_genes_n = tf_go_df[[1]]$hgnc_symbol %>% unique %>% length,
                tf_genes_frac = (tf_genes_n / group_markers_n) %>% round(2),
                tf_go_df = tf_go_df
          )
  
  tf_cofactors <- tibble(# group = "rg_pfc_all", 
                         
                         tf_cofactor_genes = tf_cofactor_go_df[[1]]$hgnc_symbol %>% 
                                                  unique %>% paste(collapse= ";"),
                         
                         tf_cofactor_genes_n = tf_cofactor_go_df[[1]]$hgnc_symbol %>% 
                                                  unique %>% length,
                         tf_cofactor_genes_frac = (tf_cofactor_genes_n / group_markers_n) %>% round(2),
                         
                         tf_cofactor_go_df = tf_cofactor_go_df
                       
                  )
  
  tfs_and_cofactors <- bind_cols(tfs, tf_cofactors)
  
}

 %>% left_join(expression_go) %>% add_column(all_markers_go_df = list(goids_x))
```

# Repeat for all RG groups in `pairs$by_area` list.

```{r}
groups_byArea <- pairs$by_area %>% filter(item2 %>% str_detect("_all")) %>% ungroup

rg_go_df <- getGO_map(groups_byArea, "rg")
neuron_go_df <- getGO_map("neuron")
ipc_go_df <- getGO_map("ipc")

go_dfs <- bind_rows(rg_go_df, neuron_go_df, ipc_go_df)
```

# Join functions above

```{r}
go_df_expression <- go_dfs$go_df %>% map(~ getGenesWithExpressionGO(go_df = .x)) %>% reduce(bind_rows)

tfs_and_cofactors <- go_dfs$go_df %>% map(~ getGenesWithTFannotation(go_df = .x)) %>% reduce(bind_rows)

go_tfs_cofactors_df <- bind_cols(go_dfs, tfs_and_cofactors, go_df_expression)

go_tfs_cofactors_df %>% select(- ends_with("_df")) %>%
write_tsv("../transcription_factors/markers_by_area_go_tfs_cofactors_df.tsv")
```

# Plots . . . . . . . . . . . . . . . . . . . . . . . . . . . .

```{r}
pairs$by_area %<>% rename(celltype_target = "cell_targetpe_target")

colors_area <- read_tsv("../../general/R/colors_area.tsv")

colors_area$category <- factor(colors_area$category, levels = go_tfs_cofactors_df$region_target %>% droplevels %>% levels)

colors_area <- colors_area %>% arrange(category)

```

```{r}
makeBarPlots <- function(go_tf_df, pairs_tbl, stage = NULL) {
  
  go_tf_df %<>% left_join(pairs_tbl %>% ungroup %>% select(group = "item2", ends_with('target'))) %>% 
  arrange(region_target, celltype_target) %>%
  mutate(group = as_factor(group))
  
  p1 <- go_tf_df %>% 
  
  ggplot +   ggtitle("All marker genes") +
  geom_col(aes(y = group, x = group_markers_n, 
               fill = region_target, alpha = factor(celltype_target))) +
    
  scale_fill_manual(values = colors_area$color.region[1:4], guide = "none") +
  scale_alpha_discrete(range = c(0.5, 0.9), guide = "none") +
  scale_x_continuous(limits = c(0, 900)) +

  theme_minimal()


p2 <- go_tf_df %>% 
  
  ggplot + ggtitle("TF genes\n / all group markers") +
  geom_col(aes(y = group, x = tf_genes_frac, 
               fill = region_target, alpha = factor(celltype_target))) +
  
  scale_fill_manual(values = colors_area$color.region[1:4], guide = "none") +
  scale_alpha_discrete(range = c(0.5, 0.9), guide = "none") +
  scale_x_continuous(limits = c(0, 0.23)) +
  
  theme_minimal() +
  theme(axis.text.y = element_blank(),
        axis.title.y = element_blank(),
        axis.ticks.y = element_blank()
        ) 


p3 <- go_tf_df %>% 
  
  ggplot +   ggtitle("Transcription co-factor genes\n / all group markers") +
  geom_col(aes(y = group, x = tf_cofactor_genes_frac, 
               fill = region_target, alpha = factor(celltype_target))) +
  
  scale_fill_manual(values = colors_area$color.region[1:4], guide = "none") +
  scale_alpha_discrete(range = c(0.5, 0.9), guide = "none") +
  scale_x_continuous(limits = c(0, 0.23)) +

  theme_minimal() +
  theme(axis.text.y = element_blank(),
        axis.title.y = element_blank(),
        axis.ticks.y = element_blank()
        ) 



p4 <- go_tf_df %>% 
  
  ggplot + ggtitle("Genes w/ expression-related GO\n / all group markers") +
  geom_col(aes(y = group, x = expression_go_genes_frac, 
               fill = region_target, alpha = factor(celltype_target))) +
  
  scale_fill_manual(values = colors_area$color.region[1:4], guide = "none") +
  scale_alpha_discrete(range = c(0.5, 0.9), guide = "none") +
  
  theme_minimal() +
  theme(axis.text.y = element_blank(),
        axis.title.y = element_blank(),
        axis.ticks.y = element_blank()
        ) 

plots <- (p1 + p2 + p3 + p4) +
  plot_layout(
    ncol = 4,
    guides = "collect"
    ) 
  plot_annotation(
  title = paste('Areal marker genes by cell type:', stage, 'second trimester')
  )

ggsave(plot = plots, 
       filename = paste0("./out/fractionGenes_barchart_tf_cofactor_expressionGO_byStage_", stage, ".pdf"), 
       width = 13, height = 8)

return(plots)

}
```

```{r}

ggsave(plot = plots, filename = "./out/tf_cofactor_expressionGO_fractionGenes_barchart.pdf", width = 13, heigh = 8)

# TODO: remove legend redundancy and strip '_all' 
# add counts at tip of barplot?
# Fix / or remove axis titles
```

# Do this for the different stages

```{r}
groups_byArea_stage <- pairs$by_area_stage %>% filter(item2 %>% str_detect("_all")) %>% ungroup

tfs_gos_byStage <- getTFsAndGOs(groups_byArea_stage)

# Prepare TFs df for dotPlots
factor_order <- list(
  celltype = as_factor(c('rg', 'ipc', 'neuron')),
  region = as_factor(c('pfc', 'msp', 'temporal', 'v1')),
  stage = as_factor(c('early', 'mid', 'late'))
)

# Reorder tf_gos_byStage group levels
tfs_gos_byStage %<>% 
  separate(group, into = c('celltype', 'region', 'stage'), 
            remove = FALSE, sep = "_") %>%
    mutate(celltype = factor(celltype, levels = factor_order$celltype),
            region = factor(region, levels = factor_order$region),
            stage = factor(stage, levels = factor_order$stage)) %>% 
   arrange(celltype, stage, region)


tfs_gos_byStage %>% select(- ends_with("_df")) %>%
write_tsv("../out/markers_by_area_and_stage_go_tfs_cofactors_df.tsv")

pairs$by_area_stage %<>% rename(celltype_target = "cell_targetpe_target")

makeBarPlots(go_tf_df = tfs_gos_byStage %>% filter(group %>% str_detect('early')), pairs_tbl = pairs$by_area_stage)

stages <- c("early", "mid", "late")

plots_by_stages <- map(stages, ~ makeBarPlots(go_tf_df = tfs_gos_byStage %>% 
                             filter(group %>% str_detect(.x)), 
                           pairs_tbl = pairs$by_area_stage,
                           stage = .x)
                    )
wrap_plots(plots_by_stages,nrow = 3)

ggsave(plot = wrap_plots(plots_by_stages,nrow = 3), 
       filename = "./out/tf_cofactor_expressionGO_fractionGenes_barchart_byStages.pdf", width = 13, height = 8)
```

```{r}
getTFsAndGOs <- function(groups_tbl) {
  
  # go_dfs_tbl  <- getGO_map(groups_tbl = groups_tbl)

  tfs_and_cofactors <- go_dfs_tbl$go_df %>% map(~ getGenesWithTFannotation(go_df = .x)) %>% reduce(bind_rows)
  
  go_df_expression <- go_dfs_tbl$go_df %>% map(~ getGenesWithExpressionGO(go_df = .x)) %>% reduce(bind_rows)

  go_tfs_cofactors_df <- bind_cols(go_dfs_tbl, tfs_and_cofactors, go_df_expression)

}
```

```{r}
#TODO [2020-11-13]
# [x] Arrange 'region' as factor
# [] Join with gene scores ? Combine gene scores of different ages into their stage.
# [] Make violin plots to confirm?
# [] Add gw16 to ncx.exn


# https://stackoverflow.com/questions/48623753/extract-a-dataframe-from-a-column-of-dataframes-tidyverse-approach
genes <- go_tfs_cofactors_df %>% filter(group == "rg_pfc_all") %>% 
              select(tf_go_df) %>% unnest %>% pull(hgnc_symbol) %>% unique
```

```{r 2020-11-25_dotPlot}

Idents(ncx.exn) <- 'celltype_?'
Idents(ncx.exn) %>% table

dotplot <- DotPlot(# dot.min = 0.01,
                   object = ncx.exn,
                  idents = c('rg_early', 'rg_mid', 'rg_late'),
                  features = genes,
                          group.by = 'region_stage',
                          scale = TRUE) + 
  theme(axis.text.x = element_text(angle = 45)) 

dotplot %>% ggplotly()
                          # dot.scale = 1)

[Id4 is required for the correct timing of neural differentiation - ScienceDirect]
(https://www.sciencedirect.com/science/article/pii/S0012160605000734)


ggsave(plot = dotplot, "out/plots/dotplot_rg_early.pdf", width = 20, height = 3)
```

## Violin Plots

```{r}
source_rmd(file = "~/cse-phd/second-trimester/R/stackedViolinPlots.Rmd")

StackedVlnPlot(ncx.exn, features = genes[1:2], title = "RG PFC TFs", 
               idents = c('rg_early', 'rg_mid', 'rg_late'), 
               group.by = 'region_stage', 
               split.by = 'region')
```

```{r}
  distinct(hgnc_symbol, .keep_all = TRUE) %>% dplyr::select(hgnc_symbol, name_1006, n, everything())
# 149 genes are classified as some sort of transcription GO
%>% 
  ggplot() + geom_tile(aes(y = name_1006, x = as_factor(hgnc_symbol), fill = name_1006)) + theme(legend.position = "none", axis.text.x = element_text(angle = 90))
```

```{r}

goids_x %>% inspectdf::inspect_cat()
```

```{r}
goids_x %>% dplyr::select(namespace_1003, name_1006) %>% ggplot() + geom_bar(aes(y = name_1006)) + facet_wrap(~namespace_1003, scales = 'free')
```
